### PR TITLE
Compound sql parse error

### LINKED ISSUE #894: Complex sql conditions break EM-training if not in conjunctive normal form
### What happens?

If you have, within some comparison level, some logical formula that is not expressed in [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) then the EM training step may break.

This is due to the fact that without passing `comparison_levels_to_reverse_blocking_rule` means they are computed with `Settings._get_comparison_levels_corresponding_to_training_blocking_rule`, ultimately using `ComparisonLevel._is_exact_match`, which splits the condition on `AND`s, parsing each bit of sql separately. This can then fail if there are any `AND`s nested below the top-layer of the condition, for instance in an expression such as `(A AND B) OR (C AND D)`. The `sqlglot` parser will see an expression `(A ` and fail due to lack of closing parenthesis: `sqlglot.errors.ParseError: Expecting ).`

This can be worked around by passing `comparison_levels_to_reverse_blocking_rule` (which comes with some mental overhead), or by rewriting the sql statement in conjunctive normal form, i.e. so that it is a series of smaller, `AND`-less clauses combined together with `AND`s (which could be complicated to derive in practice, and may hurt readability).

### To Reproduce

```py
import pandas as pd
import numpy as np

from splink.duckdb.duckdb_linker import DuckDBLinker
import splink.duckdb.duckdb_comparison_library as cl
import splink.duckdb.duckdb_comparison_level_library as cll
from splink import __version__

print(__version__)

np.random.seed(11)

companies = (
    "Wilbrinkingtons",
    "Green Camel Solutions",
    "Inverted Chessboard Industries",
    "Davis and Vadis",
    pd.NA,
)
names = (
    "Grolter",
    "Janklew",
    "Hartilty",
    "Ploblibb",
    "Swanglini",
    pd.NA
)

n_rows = 2000
df = pd.DataFrame(
    {
        "id": range(0, n_rows),
        "name": np.array(list(map(lambda x: np.random.choice(names, 1)[0], range(0, n_rows)))),
        "company_1": np.array(list(map(lambda x: np.random.choice(companies, 1)[0], range(0, n_rows)))),
        "company_2": np.array(list(map(lambda x: np.random.choice(companies, 1)[0], range(0, n_rows)))),
        "company_3": np.array(list(map(lambda x: np.random.choice(companies, 1)[0], range(0, n_rows)))),
    }
)
print(df.head())

def _col_is_null(col):
    return f"({col}_l IS NULL OR {col}_r IS NULL)"

sql_two_out_of_three_null = (
    f"(({_col_is_null('company_1')} AND {_col_is_null('company_2')}) OR "
    f"({_col_is_null('company_2')} AND {_col_is_null('company_3')}) OR "
    f"({_col_is_null('company_3')} AND {_col_is_null('company_1')}))"
)

settings = {
    "unique_id_column_name": "id",
    "link_type": "dedupe_only",
    "comparisons": [
        cl.exact_match("name"),
        {
            "output_column_name": "company_comparison",
            "comparison_levels": [
                {
                    "sql_condition": f"{_col_is_null('company_1')} AND {_col_is_null('company_2')} AND {_col_is_null('company_3')}",
                    "label_for_charts": "NULL",
                    "is_null_level": True,
                },
                {
                    "sql_condition": sql_two_out_of_three_null,
                    "label_for_charts": "2/3 NULL",
                },
                cll.exact_match_level("company_1"),
                cll.exact_match_level("company_2"),
                cll.exact_match_level("company_3"),
                cll.else_level(),
            ],
        }
    ]
}

linker = DuckDBLinker(df, settings)

linker.estimate_parameters_using_expectation_maximisation(
    "l.name = r.name"
)
```

### OS:

Mac OS

### Splink version:

3.4.2

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by ADBond**
Solves issue where compound expressions in comparison levels could break EM-training, by transforming SQL to [conjunctive normal form](https://en.wikipedia.org/wiki/Conjunctive_normal_form). Internal functions that determine exact-match status now work with `sqlglot` ASTs directly rather than raw SQL, to avoid having to parse multiple times.

Big thanks to @tobymao for pointing me in the direction of `sqlglot`'s `normalise` function to transform SQL into conjunctive normal form, which is doing the heavy lifting here.

Fixes #894.

### CODE REVIEW

File: tests/test_compound_comparison_levels.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        ]
+    )
+
+    def col_is_match(col):
+        return f"({col}_l = {col}_r)"
+
+    def col_is_null(col):
+        return f"({col}_l IS NULL OR {col}_r IS NULL)"
+
+    sql_two_out_of_three_match = (
```
Comment (NickCrews) [inline tests/test_compound_comparison_levels.py:77]: Rename this variable to `sql_ands_joined_with_ors` or `sql_non_CNF` or something to make it more clear what this is testing.

**Comment by NickCrews** [comment]
Looks great to me @ADBond, thanks so much for doing this!

File: splink/comparison_level.py [hunk 2]
```diff
@@ -68,6 +69,15 @@ def _exact_match_colname(sql, sql_dialect=None):
     return cols[0]
 
 
+def _get_and_subclauses(expr):
```
Comment (ThomasHepworth) [inline splink/comparison_level.py:72]:
```suggestion
def _get_and_subclauses(expr: sqlglot.Expression):
```

Comment (ThomasHepworth) [reply splink/comparison_level.py:72]: ☝️ for clarity. Calling the argument `expr` is fine, but slightly ambiguous.

File: tests/test_compound_comparison_levels.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        "col_6_l = col_6_r",
+        "col_7_l = col_7_r",
+    )
+
+    complex_condition_sql = " OR ".join(
+        [
+            f"({A} AND {B})",
+            f"NOT ({C} AND {D})",
+            f"({B} AND NOT {E})",
+            f"({F} AND (NOT {A} AND {G}))",
```
Comment (ThomasHepworth) [inline tests/test_compound_comparison_levels.py:195]: You could also check the new sqlglot `normlize` function is doing its job as expected here.

`A and B OR C` -> `(a OR c) AND (b OR c)`

**Review by ThomasHepworth — APPROVED** [review]

**Pull request merged**